File: src/validator.rs
use crate::types::{AnimationClipData, EasingType, InstanceData};

pub struct Validator;

impl Validator {
    pub fn validate_clip(clip: &AnimationClipData) -> Result<(), String> {
        if clip.id.is_empty() {
            return Err("Clip ID cannot be empty".to_string());
        }
        if clip.duration < 0.0 {
            return Err(format!("Clip '{}' duration cannot be negative ({})", clip.id, clip.duration));
        }

        let mut prev_time = -1.0;
        for (idx, kf) in clip.keyframes.iter().enumerate() {
            if kf.time < 0.0 {
                return Err(format!(
                    "Clip '{}' keyframe #{} time cannot be negative ({})",
                    clip.id, idx, kf.time
                ));
            }
            if kf.time > clip.duration {
                return Err(format!(
                    "Clip '{}' keyframe #{} time ({}) exceeds clip duration ({})",
                    clip.id, idx, kf.time, clip.duration
                ));
            }
            if kf.time < prev_time {
                return Err(format!(
                    "Clip '{}' keyframes must be monotonically non-decreasing in time: keyframe #{} ({}) < previous ({})",
                    clip.id, idx, kf.time, prev_time
                ));
            }
            prev_time = kf.time;

            if kf.easing == EasingType::CubicBezier {
                if let Some(params) = &kf.cubic_params {
                    if params.p1x < 0.0 || params.p1x > 1.0 {
                        return Err(format!(
                            "Clip '{}' keyframe #{} cubic bezier p1x must be in [0, 1], got {}",
                            clip.id, idx, params.p1x
                        ));
                    }
                    if params.p2x < 0.0 || params.p2x > 1.0 {
                        return Err(format!(
                            "Clip '{}' keyframe #{} cubic bezier p2x must be in [0, 1], got {}",
                            clip.id, idx, params.p2x
                        ));
                    }
                }
            }
        }

        Ok(())
    }

    pub fn validate_instance(inst: &InstanceData) -> Result<(), String> {
        if inst.id.is_empty() {
            return Err("Instance ID cannot be empty".to_string());
        }
        if inst.clip_id.is_empty() {
            return Err(format!("Instance '{}' must specify a clip_id", inst.id));
        }
        if inst.time_remapping_speed <= 0.0 {
            return Err(format!(
                "Instance '{}' time_remapping_speed must be positive, got {}",
                inst.id, inst.time_remapping_speed
            ));
        }
        Ok(())
    }

    pub fn validate_clip_references(
        instances: &[InstanceData],
        clips: &[AnimationClipData],
    ) -> Result<(), String> {
        let mut seen_instance_ids = std::collections::HashSet::new();

        for inst in instances {
            let found = clips.iter().any(|c| c.id == inst.clip_id);
            if !found {
                return Err(format!(
                    "Instance '{}' references clip_id '{}' which does not exist",
                    inst.id, inst.clip_id
                ));
            }
            if let Some(info) = &inst.inherit_from {
                let src_found = seen_instance_ids.contains(&info.source_instance_id);
                if !src_found {
                    let exists_later = instances.iter().any(|i| i.id == info.source_instance_id);
                    if exists_later {
                        return Err(format!(
                            "Topological order violation: Inheriting instance '{}' references source_instance_id '{}' which appears after '{}' in instances array",
                            inst.id, info.source_instance_id, inst.id
                        ));
                    } else {
                        return Err(format!(
                            "Instance '{}' inherits from source_instance_id '{}' which does not exist",
                            inst.id, info.source_instance_id
                        ));
                    }
                }
            }
            seen_instance_ids.insert(inst.id.clone());
        }
        Ok(())
    }
}

File: js/validator.ts
import type { Clip } from './builder/clip.js';
import type { Instance } from './builder/instance.js';
import type { AnimationClipData, InstanceData } from './builder/types.js';
import { Easing } from './builder/types.js';

export class Validator {
  static validateClip(clipInput: Clip | AnimationClipData): { ok: boolean; error?: string } {
    const clip = typeof (clipInput as any).build === 'function' ? (clipInput as Clip).build() : (clipInput as AnimationClipData);

    if (!clip.id || !clip.id.trim()) {
      return { ok: false, error: 'Clip ID cannot be empty' };
    }
    if (clip.duration < 0) {
      return { ok: false, error: `Clip '${clip.id}' duration cannot be negative (${clip.duration})` };
    }

    let prevTime = -1;
    const keyframes = clip.keyframes || [];
    for (let idx = 0; idx < keyframes.length; idx++) {
      const kf = keyframes[idx];
      if (kf.time < 0) {
        return { ok: false, error: `Clip '${clip.id}' keyframe #${idx} time cannot be negative (${kf.time})` };
      }
      if (kf.time > clip.duration) {
        return { ok: false, error: `Clip '${clip.id}' keyframe #${idx} time (${kf.time}) exceeds duration (${clip.duration})` };
      }
      if (kf.time < prevTime) {
        return { ok: false, error: `Clip '${clip.id}' keyframes must be monotonically non-decreasing in time: keyframe #${idx} (${kf.time}) < previous (${prevTime})` };
      }
      prevTime = kf.time;

      if (kf.easing === Easing.CubicBezier && kf.cubic_params) {
        const { p1x, p2x } = kf.cubic_params;
        if (p1x < 0 || p1x > 1) {
          return { ok: false, error: `Clip '${clip.id}' keyframe #${idx} cubic bezier p1x must be in [0, 1], got ${p1x}` };
        }
        if (p2x < 0 || p2x > 1) {
          return { ok: false, error: `Clip '${clip.id}' keyframe #${idx} cubic bezier p2x must be in [0, 1], got ${p2x}` };
        }
      }
    }

    return { ok: true };
  }

  static validateInstance(instInput: Instance | InstanceData): { ok: boolean; error?: string } {
    const inst = typeof (instInput as any).build === 'function' ? (instInput as Instance).build() : (instInput as InstanceData);

    if (!inst.id || !inst.id.trim()) {
      return { ok: false, error: 'Instance ID cannot be empty' };
    }
    if (!inst.clip_id || !inst.clip_id.trim()) {
      return { ok: false, error: `Instance '${inst.id}' must specify a clip_id` };
    }
    if (inst.time_remapping_speed !== undefined && inst.time_remapping_speed <= 0) {
      return { ok: false, error: `Instance '${inst.id}' time remapping speed must be positive, got ${inst.time_remapping_speed}` };
    }
    return { ok: true };
  }

  static validateReferences(instancesInput: (Instance | InstanceData)[], clipsInput: (Clip | AnimationClipData)[]): { ok: boolean; error?: string } {
    const clips = clipsInput.map((c) => typeof (c as any).build === 'function' ? (c as Clip).build() : (c as AnimationClipData));
    const instances = instancesInput.map((i) => typeof (i as any).build === 'function' ? (i as Instance).build() : (i as InstanceData));

    const clipIds = new Set(clips.map((c) => c.id));
    for (const inst of instances) {
      if (!clipIds.has(inst.clip_id)) {
        return { ok: false, error: `Instance '${inst.id}' references clip_id '${inst.clip_id}' which does not exist` };
      }
    }
    return { ok: true };
  }
}
